discovery/openstack/instance.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	ticker := time.NewTicker(i.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := i.refresh()
+			if err != nil {
+				i.logger.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (i *InstanceDiscovery) refresh() (*config.TargetGroup, error) {
+	var err error
+	t0 := time.Now()
+	defer func() {
+		refreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			refreshFailuresCount.Inc()
+		}
+	}()
+
+	provider, err := openstack.AuthenticatedClient(*i.authOpts)
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack session: %s", err)
+	}
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: i.region,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack compute session: %s", err)
+	}
+
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+
+	tg := &config.TargetGroup{
+		Source: fmt.Sprintf("OS_" + i.region),
+	}
+
+	pagerFIP := floatingips.List(client)
+	floatingIPList := make(map[string][]string)
+
+	err = pagerFIP.EachPage(func(page pagination.Page) (bool, error) {
+		result, err := floatingips.ExtractFloatingIPs(page)
+		if err != nil {
+			i.logger.Warn(err)
+		}
+		for _, ip := range result {
+			// Skip not associated ips
+			if ip.InstanceID != "" {
+				floatingIPList[ip.InstanceID] = append(floatingIPList[ip.InstanceID], ip.IP)
+			}
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("could not describe floating IPs: %s", err)
+	}
+
+	err = pager.EachPage(func(page pagination.Page) (bool, error) {
+		serverList, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, fmt.Errorf("could not extract servers: %s", err)
+		}
+
+		for _, s := range serverList {
+			labels := model.LabelSet{
+				openstackLabelInstanceID: model.LabelValue(s.ID),
+			}
+			if len(s.Addresses) == 0 {
+				i.logger.Info("Got no IP address for instance %s", s.ID)
+				continue
+			}
+
+			for _, address := range s.Addresses {
+				md, ok := address.([]interface{})
+				if !ok {
+					i.logger.Warn("Invalid type for address, expected array")
+					continue
+				}
+
+				if len(md) == 0 {
+					i.logger.Debugf("Got no IP address for instance %s", s.ID)
+					continue
+				}
+
+				md1, ok := md[0].(map[string]interface{})
+				if !ok {
+					i.logger.Warn("Invalid type for address, expected dict")
+					continue
+				}
+
+				addr, ok := md1["addr"].(string)
+				if !ok {
+					i.logger.Warn("Invalid type for address, expected string")
+					continue
+				}
+
+				labels[openstackLabelPrivateIP] = model.LabelValue(addr)
+
+				addr = net.JoinHostPort(addr, fmt.Sprintf("%d", i.port))
+
+				labels[model.AddressLabel] = model.LabelValue(addr)
+
+				// Only use first private IP
+				break
+			}
+
+			if val, ok := floatingIPList[s.ID]; ok {
+				if len(val) > 0 {
```

Comment:
No need for another if and indentation: `if val, ok := floatingIPList[s.ID]; ok && len(val) > 0{`

Reply:
Updated.